### ISSUE TITLE
Cluster metadata operation not returning topics

### ISSUE THREAD

**Issue by bsamartins** [CallistoLabsNYC/samsa#103]
Cluster metadata operation does not return topics when none is provided.

```rust
use samsa::prelude::{ClusterMetadata, TcpConnection};
use std::fmt::Error;
use tracing::level_filters::LevelFilter;
use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;

#[tokio::main]
async fn main() -> Result<(), Error> {
    // Start logging to console
    tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::builder()
                .with_default_directive(LevelFilter::TRACE.into())
                .from_env_lossy()
        )
        .with(tracing_subscriber::fmt::Layer::default().compact())
        .init();
    
    let bootstrap_addrs = vec![samsa::prelude::BrokerAddress {
        host: "localhost".to_owned(),
        port: 9092,
    }];

    let topics = match ClusterMetadata::<TcpConnection>::new(
        bootstrap_addrs, 
        0, 
        "test".to_owned(), 
        vec![],
     ).await {
        Ok(metadata) => metadata.topic_names,
        Err(err) => { panic!("Failed to retrieve cluster metadata: {}", err); }
    };

    println!("{:?}", topics);

    Ok(())
}
```

**Comment by morukele** [comment]
Hello @bsamartins, I am trying to investigate this issue. From my understanding of the source code, the `ClusterMetadata` struct is constructed with a vec of empty topic and is populated with the topic passed into its constructor. 

I think that this behaviour is expected as the `topics` variable is an empty vec, the same that was passed in. Could you shed more light on what you are trying to achieve?

**Comment by bsamartins** [comment]
@morukele , I'm just trying to list all the broker's topics.
When I executed the request, I got none back.

**Comment by dhonig** [comment]
@morukele  I think @bsamartins  has a reasonable expectation and therefore we should include this feature.
From this doc: https://kafka.apache.org/protocol.html
'''
How can the client find out which topics exist, what partitions they have, and which brokers currently host those partitions so that it can direct its requests to the right hosts? This information is dynamic, so you can't just configure each client with some static mapping file. Instead all Kafka brokers can answer a metadata request that describes the current state of the cluster: what topics there are, which partitions those topics have, which broker is the leader for those partitions, and the host and port information for these brokers.

In other words, the client needs to somehow find one broker and that broker will tell the client about all the other brokers that exist and what partitions they host. This first broker may itself go down so the best practice for a client implementation is to take a list of two or three URLs to bootstrap from. The user can then choose to use a load balancer or just statically configure two or three of their Kafka hosts in the clients.
'''
It seems this is what @bsamartins is expecting.   We should provide this functionality in a manner that complies with the description in the protocol guide.   That's awesome if you want to take this on.  @hgm-king  will be here to support you.

**Comment by morukele** [comment]
@hgm-king I will review the protocol document and add an implementation for the feature. I am conflicted on whether it should be a separate function in the ClusterMetaData struct or if it should just be added to the topics field of the struct directly.

**Comment by daniel-honig-clearscale** [comment]
@morukele I wouldn't worry too much about the question above. I think ClusterMetaData struct is fine to start and we can review it together and move it later.

**Comment by dhonig** [comment]
@morukele I wouldn't worry too much about the question above. I think ClusterMetaData struct is fine to start and we can review it together and move it later.

**Comment by dhonig** [comment]
@morukele any updates on this?

**Comment by morukele** [comment]
@dhonig, I am still working on it; I hope to have an update by the end of this week. I have had a busy week.

**Comment by morukele** [comment]
Hello @dhonig and @bsamartins, I have addressed this issue in this pull request #119.

If merged, the code above should work
